### ISSUE TITLE
Clarify that normative behavior for `collections.abc.Mapping.__iter__` is that of `dict.__iter__()`

### ISSUE THREAD

**Issue by eirrgang** [python/cpython#98760]
# Documentation

## Summary

It seems like it is the intent that a `Mapping` (`typing.Mapping` or `collections.abc.Mapping`) should yield its keys when iterated over. This intent should be clearer.

https://docs.python.org/3/library/stdtypes.html#dictionary-view-objects is too hard to find, and does not explicitly describe the expected behavior of `collections.abc.MappingView` and `collections.abc.Mapping`

## Notes

Starting from https://docs.python.org/3/library/typing.html#typing.Mapping , a reader can follow roughly the path of
1. https://docs.python.org/3/library/collections.abc.html
2. https://docs.python.org/3/library/collections.abc.html#collections.abc.Mapping 
3. https://docs.python.org/3/glossary.html#term-dictionary-view
4. https://docs.python.org/3/library/stdtypes.html#dict-views
5. "iter(dictview): Return an iterator over the keys..."

Wrong turns along the path could terminate at
* https://docs.python.org/3/library/stdtypes.html#iterator.__iter__
* https://docs.python.org/3/glossary.html#term-mapping

`typeshed` has a `types.pyi` that defines `def __iter__(self) -> Iterator[_KT]: ...` for `MappingProxyType`, but `typing.pyi` misses the opportunity to override the prototype for `Mapping` or `MappingView`

## Possible improvements

Clarify in `glossary` that dictionary views are normative behavior for Mapping views.

Note the `def __iter__(self) -> Iterator[_KT]: ...` signature (in prose or syntax) in https://docs.python.org/3/library/collections.abc.html

<s>Update the typeshed `Mapping` abc with the correct `__iter__` signature.</s> *edit: would it be inappropriate to update the `collections.abc.Mapping` definition? I'm not sure how mypy ends up applying the `MappingProxyType.__iter__` signature to `Mapping`, but it seems to do so. However, at least some other static type checkers do not. If bug reports should be filed for these, it would help to affirm in this issue what their behavior should be and how that behavior should be inferred. Also, I was under the impression that the full Generic abc should be clear from the `typing` (and now `collections.abc`) module(s) without reliance on the typeshed pyi files, such that subclassing from `collections.abc` ABCs should be sufficient guidance at authorship time without reliance on a separate type checking tool. However, I have seen mentioned that Generic syntax is intentionally avoided in the stdlib module files, so maybe it is inappropriate to assume that type signatures for overloads should be obvious solely through the mro, inspecting of (virtual) super class definitions, or actual values of annotations.*

**Comment by eirrgang** [comment]
I would be happy to submit one or more patches, if someone could weigh in on what directions are or aren't appropriate.

**Comment by rhettinger** [comment]
Sorry, I don't a real problem.  By the time a user gets to a Mapping ABC, they are already well experienced with. dictionaries (which are introduced early in the tutorial).   I've never seen. users confused by this. Thanks for the suggestion, but I don't think there is an actual problem to be solved.  

Also, the notation of "normative behaviors" itself a bit confusing, creating a tension between what can be done versus what is commonly done.  The docs focus on the former except in cases we there is common and recurring confusion

**Comment by eirrgang** [comment]
Maybe it's less of an issue for new readers of the documentation. Or maybe
I'm just not as clever as i thought i was.

The potentially embarrassing personal case:
I had various classes (originally written for Py 2.7) that at some point
came to overrode both __getitem__ and __iter__. Some of them evolved and
came to use collections.abc.Mapping or typing.Mapping when they appeared,
because it seemed appropriate. When I recently switched to Generic
versions, my classes did _not_ fail a static type check, though it seems
like they should have, and it seems like it took me an embarrassingly long
time to discover an implication that Mapping was assumed to iter like a
dict.

If Mapping is strictly meant to be a Generic Dict, with possibly unexpected
behaviors like an iterator that produces keys, it wasn't clear enough to
me. If this is not the intention, the non-normative nature (of dict iter
behavior for all Mapping) is contradicted by the type hint I mentioned.

**Comment by eirrgang** [comment]
If the there is not an intention that iter on Mapping yields keys, then it
seems like the resolution to this issue would be to relax the typevar in
the MappingProxyType.__iter__ type hint (so dummies like me don't suddenly
think we've been doing it "wrong" all these years ;-) ).

**Comment by rhettinger** [comment]
> When I recently switched to Generic
versions, my classes did _not_ fail a static type check, though it seems
like they should have, and it seems like it took me an embarrassingly long
time to discover an implication that Mapping was assumed to iter like a dict.

I am not able to reproduce your static type check issue.  In the following code, MyPy correctly identifies signature errors in both `__getitem__` and `__iter__`:

```
from typing import Iterator
from collections.abc import Mapping

class M(Mapping[str, int]):

    def __iter__(self) -> Iterator[bytes]:
        yield b'a'
        yield b'b'

    def __getitem__(self, key: str) -> int:
        return ord(key[0])

    def __len__(self):
        return 2
```

In what ways, did your code differ?

**Comment by eirrgang** [comment]
Yes, with mypy 0.982, I get an error that `Iterator[bytes]` does not match `Iterator[str]`. (If I remove the return type annotation, I do not get an error. I'm a bit fuzzy on the expectations for type hint inheritance and overloads, so I don't know if this is appropriate.)

I think that this Generic behavior is defined in typeshed, but not through `typing` or `collections.abc`. This leads to confusing feedback to a programmer who is relying on the ABC mro to correctly implement/subclass `Mapping[KT, VT]`.

If the example you give is supposed to produce an error, then I interpret that to mean that `Mapping[KT, VT].__iter__(self) -> Iterator[KT]` is "normative". I hope that this information could be made more discoverable and more obvious through the `collections.abc.Mapping` definition. *

If `Mapping[KT, VT].__iter__(self) -> Iterator[KT]` is "not normative", and `Mapping[KT, VT].__iter__(self) -> Iterable[_T_co]` is reasonable (as the py 3.9 and py 3.10 mro imply to me), then the typeshed spec for `MappingProxyType.__iter__` is too restrictive.

* It may also imply that bug reports should be filed against some static type checkers, though I'm not sure why the `MappingProxyType.__iter__` return type annotation should necessarily be applied to `Mapping.__iter__`. I don't remember exactly what brought this inconsistency to my attention, but it may have been that I was using pylance or something with editor integrations and then noticed different results from mypy.

**Comment by rhettinger** [comment]
A few thoughts:

* MyPy and typeshed seem to be working as intended. For me, it reliably detected a type mismatch for `__iter__`.

* The Mapping ABC is also working as intended.  It fulfilled its limited goals 1) verify `__iter__` is present prior to instantiation, 2) provide mixin methods that call `__iter__`, and 3) satisfy isinstance/issubclass checks.

* The docs seem fine to me and I haven't encountered any cases of users not understanding the Mapping is dictlike, Sequence is sequencelike and Setsare setlike.  How closely a concrete class adheres to those relationships is up to the writer of the class.

* At this point, your bug is fixed. The ABCs have been around for.a long time and this hasn't come up with other users, so I think this can be safely closed.

* Also note that in newer versions of the documentation, the introductory text for the collections ABCs discuss in much more detail what the ABCs actually promise (which isn't much, it is mostly a declaration of intention accompanied by a verification that specific methods are present).

**Comment by eirrgang** [comment]
> 2) provide mixin methods that call __iter__,

Yes, actually, it was when I wondered how the mix-ins were implemented that I finally discovered that it `__iter__() -> KT` was important.

Thank you for confirming that `Mapping[KT, VT].__iter__() -> KT` is the intended typing. If I encounter static type checkers that do not correctly identify this, I can follow up with those projects, more specifically.

> Also note that in newer versions of the documentation...

I'm sorry. It looks like I didn't notice the new content in https://docs.python.org/3/library/collections.abc.html. I see that there is only a very small acknowledgement that the ABCs are subscriptable. It seems clear from that and PEP-585 that expansive type hinting is still considered beyond the scope of standard library modules and deferred to typeshed.

It also seems likely that, for Python 3.9 and beyond, the move from `typing.Mapping[]` to `collections.abc.Mapping[]` will encourage readers to consider the full range of Mapping behaviors and the implementation of the collections.abc mix-ins, which should remove any ambiguity.
